cluster/gce/util.sh

```diff
@@ -3082,13 +3151,212 @@ function create-master() {
+    --project "${PROJECT}" --region "${REGION}" -q --format='value(address)')
+  done
+}
+
+function create-tpmaster {
+  local tp_sequence=${1:-1}
+  NODE_TAG="${CLUSTER_NAME}-minion"
+  INITIAL_ETCD_CLUSTER="${MASTER_NAME}"
+  KUBERNETES_MASTER_NAME="${TPSERVER_RESERVED_IP[$tp_sequence]}"
+  MASTER_ADVERTISE_ADDRESS="${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]}"
+  KUBERNETES_MASTER_INTERNAL_IP="${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]}"
+  KUBECONFIG="${RESOURCE_DIRECTORY}/kubeconfig${KUBEMARK_PREFIX}.tp-$tp_sequence"
+
+  if [[ "${REGISTER_MASTER_KUBELET:-}" == "true" ]]; then
+    KUBELET_APISERVER="${TPSERVER_RESERVED_IP[${tp_sequence}]}"
+  fi
+
+  CERT_TEMP="${KUBE_TEMP}/tp$tp_sequence"
+  mkdir -p ${CERT_TEMP}
+  create-certs "${TPSERVER_RESERVED_IP[$tp_sequence]}" "${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]}"
+  create-etcd-certs "${TPSERVER_NAME[$tp_sequence]}" "" "" "${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]}"
+  create-etcd-apiserver-certs "etcd-${TPSERVER_NAME[$tp_sequence]}" "${TPSERVER_NAME[$tp_sequence]}" "" "" "${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]}"
+  create-scaleoutserver-instance "${TPSERVER_NAME[$tp_sequence]:-}" "${TPSERVER_RESERVED_IP[$tp_sequence]:-}" "${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]:-}"
+
+}
+
+function create-rpmaster {
+  local rp_sequence=${1:-1}
+  NODE_TAG="${CLUSTER_NAME}-minion"
+  INITIAL_ETCD_CLUSTER="${MASTER_NAME}"
+  KUBERNETES_MASTER_NAME="${RPSERVER_RESERVED_IP[${rp_sequence}]}"
+  MASTER_ADVERTISE_ADDRESS="${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  KUBERNETES_MASTER_INTERNAL_IP="${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  KUBECONFIG="${RESOURCE_DIRECTORY}/kubeconfig${KUBEMARK_PREFIX}.rp-$rp_sequence"
+
+  if [[ "${REGISTER_MASTER_KUBELET:-}" == "true" ]]; then
+    KUBELET_APISERVER="${RPSERVER_RESERVED_IP[${rp_sequence}]}"
+  fi
+
+  MASTER_EXTRA_METADATA=${KUBE_MASTER_EXTRA_METADATA:-${KUBE_EXTRA_METADATA:-}}
+  if [[ "${MASTER_EXTRA_METADATA}" != "" ]]; then
+    MASTER_EXTRA_METADATA="${MASTER_EXTRA_METADATA},${TPCONFIG_METADATA}"
+  else
+    MASTER_EXTRA_METADATA="${TPCONFIG_METADATA}"
+  fi
+
+  CERT_TEMP="${KUBE_TEMP}/rp$rp_sequence"
+  mkdir -p ${CERT_TEMP}
+  create-certs "${RPSERVER_RESERVED_IP[$rp_sequence]}" "${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  create-etcd-certs "${RPSERVER_NAME[$rp_sequence]}" "" "" "${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  create-etcd-apiserver-certs "etcd-${RPSERVER_NAME[$rp_sequence]}" "${RPSERVER_NAME[$rp_sequence]}" "" "" "${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  create-scaleoutserver-instance "${RPSERVER_NAME[$rp_sequence]:-}" "${RPSERVER_RESERVED_IP[$rp_sequence]:-}" "${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]:-}"
+
+}
+
+# master machine name format: {KUBE_GCE_ZONE}-kubemark-tp-1-master
+# destination file " --resource-providers=/etc/srv/kubernetes/kube-scheduler/rp-kubeconfig"
+# TODO: avoid calling GCE compute from here
```

Comment:
those function were originally added to kubemark and with the limitation in kubemark workflow and hence those TODOs. they probably can be removed.

Reply:
keep no change for this release